main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala

```diff
@@ -15,350 +16,170 @@
-      visitExp(exp1) ++ visitExp(exp2)
-
-    case Expr.ParYield(frags, exp, _, _, _) =>
-      frags.flatMap {
-        case ParYieldFragment(_, e, _) => visitExp(e)
-      } ++ visitExp(exp)
-
-    case Expr.Lazy(exp, _, _) =>
-      visitExp(exp)
-
-    case Expr.Force(exp, _, _, _) =>
-      visitExp(exp)
-
-    case Expr.FixpointConstraintSet(cs, _, _) =>
-      cs.flatMap(visitConstraint)
-
-    case Expr.FixpointLambda(_, exp, _, _, _) =>
-      visitExp(exp)
-
-    case Expr.FixpointMerge(exp1, exp2, _, _, _) =>
-      visitExp(exp1) ++ visitExp(exp2)
-
-    case Expr.FixpointSolve(exp, _, _, _) =>
-      visitExp(exp)
-
-    case Expr.FixpointFilter(_, exp, _, _, _) =>
-      visitExp(exp)
-
-    case Expr.FixpointInject(exp, _, _, _, _) =>
-      visitExp(exp)
+    Visitor.visitRoot(root, UseConsumer, AllAcceptor)
 
-    case Expr.FixpointProject(_, exp, _, _, _) =>
-      visitExp(exp)
+    CodeHintCandidates(traitUses, enumUses, defUses, defCalls)
+  }
 
-    case Expr.Error(_, _, _) =>
-      Nil
 
+  private def checkDeprecated(ann: Annotations, loc: SourceLocation): List[CodeHint] = {
+    if (ann.isDeprecated) { CodeHint.Deprecated(loc) :: Nil } else { Nil }
   }
 
-  /**
-    * Computes code quality hints for the given list of expressions `exps`.
-    */
-  private def visitExps(exps: List[Expr])(implicit root: Root, flix: Flix): List[CodeHint] =
-    exps.flatMap(visitExp)
-
-  /**
-    * Computes code quality hints for the given constraint `c`.
-    */
-  private def visitConstraint(c: Constraint)(implicit root: Root, flix: Flix): List[CodeHint] =
-    visitHeadPredicate(c.head) ++ c.body.flatMap(visitBodyPredicate)
+  private def checkExperimental(ann: Annotations, loc: SourceLocation): List[CodeHint] = {
+    if (ann.isExperimental) { CodeHint.Experimental(loc) :: Nil } else { Nil }
+  }
 
-  /**
-    * Computes code quality hints for the given head predicate `p`.
-    */
-  private def visitHeadPredicate(p: TypedAst.Predicate.Head)(implicit root: Root, flix: Flix): List[CodeHint] = p match {
-    case Head.Atom(_, _, terms, _, _) => visitExps(terms)
+  private def checkLazy(ann: Annotations, loc: SourceLocation): List[CodeHint] = {
+    if (ann.isLazy) { CodeHint.Lazy(loc) :: Nil } else { Nil }
   }
 
-  /**
-    * Computes code quality hints for the given body predicate `p`.
-    */
-  private def visitBodyPredicate(p: TypedAst.Predicate.Body)(implicit root: Root, flix: Flix): List[CodeHint] = p match {
-    case Body.Atom(_, _, _, _, _, _, _) => Nil
-    case Body.Functional(_, exp, _) => visitExp(exp)
-    case Body.Guard(exp, _) => visitExp(exp)
+  private def checkParallel(ann: Annotations, loc: SourceLocation): List[CodeHint] = {
+    if (ann.isParallel) { CodeHint.Parallel(loc) :: Nil } else { Nil }
   }
 
   /**
     * Checks whether `sym` would benefit from `tpe` being pure.
     */
   private def checkEffect(sym: Symbol.DefnSym, tpe: Type, loc: SourceLocation)(implicit root: Root): List[CodeHint] = {
```

Comment:
You should undo this change.

Reply:
This PR is not about changing the meaning of the phase.

Reply:
Yeah, okay. That's fair. If this indeed ought to be changed, it should be in a different PR. I get it :)